File: backend/src/main/scala/services/battle/microservices/combat/services/BattleCriticalDamageRules.scala
package services.battle.microservices.combat.services

import cats.effect.IO

import services.battle.microservices.abilities.objects.skill.SkillKind
import services.battle.microservices.actors.objects.player.BattlePlayerState
import services.battle.microservices.combat.objects.combat.Damage

private[battle] object BattleCriticalDamageRules {
  private val CriticalDamageMultiplier = 1.5

  def projectileDamage(baseDamage: Damage, owner: Option[BattlePlayerState]): IO[Damage] =
    IO.pure {
      if owner.exists(hasActiveCritical) then
        Damage(math.ceil(baseDamage.value.toDouble * CriticalDamageMultiplier).toInt)
      else baseDamage
    }

  private def hasActiveCritical(player: BattlePlayerState): Boolean =
    player.skills.exists(skill => skill.skillKind == SkillKind.Critical && skill.activeMs.value > 0L)
}

File: frontend/src/runtime/battle/microservices/combat/functions/BattleCriticalDamageRules.ts
import type { BattleHeroViewState as Hero } from "../../../../../objects/battle/microservices/actors/objects/player/BattleHeroViewState";

const CRITICAL_DAMAGE_MULTIPLIER = 1.5;

export function resolveBattleCriticalDamage(baseDamage: number, owner: Hero | null | undefined): number {
  if (!owner?.skills.some((skill) => skill.kind === "Critical" && skill.activeMs > 0)) {
    return baseDamage;
  }

  return Math.ceil(baseDamage * CRITICAL_DAMAGE_MULTIPLIER);
}
